core/lib/state/src/storage_overrides.rs

```diff
@@ -0,0 +1,149 @@
+    AccountTreeId, StorageKey, StorageValue, H256, U256,
+};
+use zksync_utils::{bytecode::hash_bytecode, h256_to_u256, u256_to_h256};
+
+use crate::{OverrideStorage, ReadStorage};
+
+/// A storage view that allows to override some of the storage values.
+#[derive(Debug)]
+pub struct StorageOverrides<S> {
+    storage_handle: S,
+    overridden_factory_deps: HashMap<H256, Vec<u8>>,
+    overridden_account_state: HashMap<AccountTreeId, HashMap<H256, H256>>,
+    overriden_account_state_diff: HashMap<AccountTreeId, HashMap<H256, H256>>,
+    overridden_balance: HashMap<StorageKey, U256>,
+    overridden_nonce: HashMap<StorageKey, U256>,
+    overridden_code: HashMap<StorageKey, H256>,
+}
+
+impl<S: ReadStorage + fmt::Debug> StorageOverrides<S> {
+    /// Creates a new storage view based on the underlying storage.
+    pub fn new(storage: S) -> Self {
+        Self {
+            storage_handle: storage,
+            overridden_factory_deps: HashMap::new(),
+            overridden_account_state: HashMap::new(),
+            overriden_account_state_diff: HashMap::new(),
+            overridden_balance: HashMap::new(),
+            overridden_nonce: HashMap::new(),
+            overridden_code: HashMap::new(),
+        }
+    }
+
+    /// Overrides a factory dependency code.
+    pub fn store_factory_dep(&mut self, hash: H256, code: Vec<u8>) {
+        self.overridden_factory_deps.insert(hash, code);
+    }
+
+    /// Overrides an account entire state.
+    pub fn override_account_state(&mut self, account: AccountTreeId, state: HashMap<H256, H256>) {
+        self.overridden_account_state.insert(account, state);
+    }
+
+    /// Overrides an account state diff.
+    pub fn override_account_state_diff(
+        &mut self,
+        account: AccountTreeId,
+        state_diff: HashMap<H256, H256>,
+    ) {
+        self.overriden_account_state_diff
+            .insert(account, state_diff);
+    }
+
+    /// Make a Rc RefCell ptr to the storage
+    pub fn to_rc_ptr(self) -> Rc<RefCell<Self>> {
+        Rc::new(RefCell::new(self))
+    }
+}
+
+impl<S: ReadStorage + fmt::Debug> ReadStorage for StorageOverrides<S> {
+    fn read_value(&mut self, key: &StorageKey) -> StorageValue {
+        if let Some(balance) = self.overridden_balance.get(key) {
+            return u256_to_h256(*balance);
+        }
+        if let Some(code) = self.overridden_code.get(key) {
+            return *code;
+        }
+
+        if let Some(nonce) = self.overridden_nonce.get(key) {
+            return u256_to_h256(*nonce);
+        }
+
+        if let Some(account_state) = self.overridden_account_state.get(key.account()) {
+            if let Some(value) = account_state.get(key.key()) {
+                return *value;
+            }
+        }
+
+        if let Some(account_state_diff) = self.overriden_account_state_diff.get(key.account()) {
+            if let Some(value) = account_state_diff.get(key.key()) {
+                return *value;
+            }
+        }
```

Comment:
For `state_diff` you have to fallback to `self.storage_handle.read_value` if key is missing